packages/mui-utils/src/useReducerCompat/useReducerCompat.ts

```diff
@@ -0,0 +1,32 @@
+import * as React from 'react';
+import {
+  runWithBatch,
+  blockingBatchCallback,
+  enqueueStateUpdate,
+} from '../CompatTransitionManager';
+
+/**
+ * Like `useReducer`, but for use with `useTransitionCompat`.
+ * TODO: Improve typing.
+ */
+function useReducerCompat<T>(reducer: any, initializerArg: any, initializer?: any) {
+  if (typeof React.startTransition === 'function') {
```

Comment:
I'm not really sure, to be honest 🫤 Consider:

```
startTransition(() => {
  dispatch('foo');
});
dispatch('bar');
```

If we use `React.useReducer` directly even in React 18, then the behavior of `dispatch` could change significantly depending the installed version of React. Specifically, state (or reducers) that would produce blocking updates in React 17 (because they don't use the shim, and we can't queue updates that don't use the shim) would suddenly produce non-blocking updates in React 18 (because React can queue them).

By using the shim even in React 18, we ensure that the behavior is always the same regardless of the React version: shimmed `useTransition` will only ever cause shimmed `useState`/`useReducer` updates to be non-blocking.

At least, those are my thoughts. If you feel particularly strongly about it though, we can change it.